src/Swarm/TUI/Controller.hs

```diff
@@ -837,32 +848,31 @@ makeEntity s e = do
           s ^? gameState . robotMap . at 0 . _Just . robotContext . defStore
 
   case isActive <$> (s ^. gameState . robotMap . at 0) of
-    Just False ->
-      continue $
-        s
-          & gameState . replStatus .~ REPLWorking mkTy Nothing
-          & gameState . robotMap . ix 0 . machine .~ initMachine mkPT empty topStore
-          & gameState %~ execState (activateRobot 0)
-    _ -> continueWithoutRedraw s
+    Just False -> do
+      gameState . replStatus .= REPLWorking mkTy Nothing
+      gameState . robotMap . ix 0 . machine .= initMachine mkPT empty topStore
+      gameState %= execState (activateRobot 0)
+    _ -> continueWithoutRedraw
 
 -- | Display a modal window with the description of an entity.
-descriptionModal :: AppState -> Entity -> EventM Name (Next AppState)
-descriptionModal s e =
-  continue $ s & uiState . uiModal ?~ generateModal s (DescriptionModal e)
+descriptionModal :: Entity -> EventM Name AppState ()
+descriptionModal e = do
+  s <- get
+  uiState . uiModal ?= generateModal s (DescriptionModal e)
```

Comment:
There are many places where we have to do `s <- get` and then call `generateModal s`.  Could we somehow rewrite `generateModal` to have a type like `(MonadReader AppState m) => generateModal :: ModalType -> m Modal` and then use some `zoom`-like thing to adapt it to `EventM Name AppState ()` ?

Reply:
To me, 2 places are not that many. :wink: It might be worth it to rewrite all functions in View to use `MonadState AppState`, but they look kind of nicely simple now.

At the moment I am leaning to leave such refactoring to other PR, it is not obvious to me how to make it nice and keep it simple.

Reply:
It looked like a lot more than 2 places to me.  In some places we do `flip generateModal`.  Anyway, leaving it for now is fine, it was just an offhanded suggestion.

Reply:
It might have been another function or maybe you looked/remembered the main branch?

I was actually looking forward to changing that function and then I came to the View module and found every function there looked like that! :sweat_smile: